File: shared/canonical.js
// shared/canonical.js — the canonical JSON form of docs/VUKA-2-SPEC.md §5.
//
// The canonical bytes are the UTF-8 of Python
//   json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
// exactly byte for byte (spec §5: "There is one JS implementation
// (shared/canonical.js), used by the app and the verify page, and one Python
// implementation in anchor/"). The rules this module enforces:
//
//   1. Object keys are ASCII only, sorted by code point, recursively.
//   2. Non-ASCII string content is escaped as \uXXXX in lowercase hex, and
//      characters outside the BMP as UTF-16 surrogate pairs (matches Python's
//      ensure_ascii=True, whose escape set is [^ -~] plus " and \).
//   3. No floats anywhere in a hashed or signed object.
//   4. Integers stay within ±(2^53 − 1).
//
// Vectors: contracts/vectors/canonical.json (T01). A rejection vector must
// fail here exactly as it fails in Python.
//
// Zero dependencies: this file is imported by the verify page and the app.

/** Raised for any value, key or JSON text the canonical form refuses. */
export class CanonicalisationError extends Error {
  constructor(message, path = "$") {
    super(`${message} (at ${path})`);
    this.name = "CanonicalisationError";
    this.path = path;
  }
}

function hex4(codeUnit) {
  return codeUnit.toString(16).padStart(4, "0");
}

// Python's ESCAPE_ASCII regex is ([\\"]|[^ -~]): printable US-ASCII (0x20–0x7E)
// stays literal, everything else escapes, DEL (0x7F) included. Iterating UTF-16
// code units encodes astral characters as lowercase surrogate pairs, which is
// exactly what Python emits, and lone surrogates pass through the same way.
function escapeString(s) {
  let out = '"';
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c === 0x22) out += '\\"';
    else if (c === 0x5c) out += "\\\\";
    else if (c === 0x08) out += "\\b";
    else if (c === 0x09) out += "\\t";
    else if (c === 0x0a) out += "\\n";
    else if (c === 0x0c) out += "\\f";
    else if (c === 0x0d) out += "\\r";
    else if (c >= 0x20 && c <= 0x7e) out += s[i];
    else out += "\\u" + hex4(c);
  }
  return out + '"';
}

function checkAsciiKey(key, path) {
  for (let i = 0; i < key.length; i++) {
    if (key.charCodeAt(i) > 0x7f) {
      throw new CanonicalisationError("object keys must be ASCII only", path);
    }
  }
}

function encodeNumber(n, path) {
  if (typeof n !== "number" || !Number.isFinite(n)) {
    throw new CanonicalisationError("only finite numbers are canonical", path);
  }
  if (!Number.isInteger(n)) {
    throw new CanonicalisationError("no floats in a hashed or signed object", path);
  }
  if (!Number.isSafeInteger(n)) {
    throw new CanonicalisationError("integers must stay within ±(2^53 − 1)", path);
  }
  // JS -0 is integer-valued; Python's integer 0 prints "0".
  return Object.is(n, -0) ? "0" : String(n);
}

function isPlainObject(v) {
  if (v === null || typeof v !== "object" || Array.isArray(v)) return false;
  const proto = Object.getPrototypeOf(v);
  return proto === Object.prototype || proto === null;
}

function encode(value, path) {
  if (value === null) return "null";
  switch (typeof value) {
    case "boolean":
      return value ? "true" : "false";
    case "number":
      return encodeNumber(value, path);
    case "string":
      return escapeString(value);
    case "object":
      break; // handled below
    default:
      // undefined, function, symbol, bigint — none of them exist in a Python
      // dict, so none of them may reach the signer.
      throw new CanonicalisationError(`type ${typeof value} is not canonical`, path);
  }
  if (Array.isArray(value)) {
    const items = value.map((item, i) => encode(item, `${path}[${i}]`));
    return "[" + items.join(",") + "]";
  }
  if (!isPlainObject(value)) {
    throw new CanonicalisationError("only plain objects are canonical", path);
  }
  const keys = Object.keys(value);
  for (const key of keys) checkAsciiKey(key, path);
  keys.sort(); // keys are ASCII, so code-unit order is code-point order
  const members = keys.map((key) => `${escapeString(key)}:${encode(value[key], `${path}.${key}`)}`);
  return "{" + members.join(",") + "}";
}

/**
 * Canonical JSON text for an in-memory value.
 * @param {unknown} value
 * @returns {string}
 */
export function canonicalJson(value) {
  return encode(value, "$");
}

/**
 * Canonical bytes (UTF-8 of the canonical text) for an in-memory value.
 * These are the bytes that go to SHA-256 or to the native signer.
 * @param {unknown} value
 * @returns {Uint8Array}
 */
export function canonicalize(value) {
  return new TextEncoder().encode(canonicalJson(value));
}
// --- strict JSON text → canonical bytes (duplicate keys and floats refused) ---

const CONTROL_CHARS = /[\u0000-\u001f]/;

class JsonParser {
  constructor(text) {
    this.text = text;
    this.pos = 0;
  }

  atEnd() {
    return this.pos >= this.text.length;
  }

  fail(message) {
    throw new CanonicalisationError(`${message} at byte ${this.pos}`);
  }

  skipWs() {
    while (!this.atEnd()) {
      const c = this.text[this.pos];
      if (c === " " || c === "\t" || c === "\n" || c === "\r") this.pos++;
      else break;
    }
  }

  expect(literal) {
    if (!this.text.startsWith(literal, this.pos)) this.fail(`expected ${literal}`);
    this.pos += literal.length;
  }

  parseValue() {
    this.skipWs();
    if (this.atEnd()) this.fail("unexpected end of JSON");
    const c = this.text[this.pos];
    if (c === "{") return this.parseObject();
    if (c === "[") return this.parseArray();
    if (c === '"') return this.parseString();
    if (c === "t") {
      this.expect("true");
      return true;
    }
    if (c === "f") {
      this.expect("false");
      return false;
    }
    if (c === "n") {
      this.expect("null");
      return null;
    }
    return this.parseNumber();
  }

  // Object keys are collected in a Set so a duplicate key is refused here,
  // before it can be collapsed. "__proto__" is defined as an own data
  // property, so a hostile payload cannot reach the prototype.
  parseObject() {
    this.expect("{");
    const obj = {};
    const seen = new Set();
    this.skipWs();
    if (this.text[this.pos] === "}") {
      this.pos++;
      return obj;
    }
    for (;;) {
      this.skipWs();
      if (this.text[this.pos] !== '"') this.fail("expected a string key");
      const key = this.parseString();
      if (seen.has(key)) {
        throw new CanonicalisationError(`duplicate key "${key}" in JSON text`);
      }
      seen.add(key);
      this.skipWs();
      if (this.text[this.pos] !== ":") this.fail("expected ':'");
      this.pos++;
      const value = this.parseValue();
      Object.defineProperty(obj, key, {
        value,
        writable: true,
        enumerable: true,
        configurable: true,
      });
      this.skipWs();
      if (this.text[this.pos] === ",") {
        this.pos++;
        continue;
      }
      if (this.text[this.pos] === "}") {
        this.pos++;
        return obj;
      }
      this.fail("expected ',' or '}'");
    }
  }

  parseArray() {
    this.expect("[");
    const arr = [];
    this.skipWs();
    if (this.text[this.pos] === "]") {
      this.pos++;
      return arr;
    }
    for (;;) {
      arr.push(this.parseValue());
      this.skipWs();
      if (this.text[this.pos] === ",") {
        this.pos++;
        continue;
      }
      if (this.text[this.pos] === "]") {
        this.pos++;
        return arr;
      }
      this.fail("expected ',' or ']'");
    }
  }
parseString() {
    // caller checked the opening quote
    this.pos++;
    let out = "";
    for (;;) {
      if (this.atEnd()) this.fail("unterminated string");
      const c = this.text[this.pos];
      if (c === '"') {
        this.pos++;
        return out;
      }
      if (c === "\\") {
        this.pos++;
        if (this.atEnd()) this.fail("unterminated escape");
        const e = this.text[this.pos++];
        switch (e) {
          case '"': out += '"'; break;
          case "\\": out += "\\"; break;
          case "/": out += "/"; break;
          case "b": out += "\b"; break;
          case "f": out += "\f"; break;
          case "n": out += "\n"; break;
          case "r": out += "\r"; break;
          case "t": out += "\t"; break;
          case "u": {
            if (this.pos + 4 > this.text.length) this.fail("incomplete \\u escape");
            const hexDigits = this.text.slice(this.pos, this.pos + 4);
            if (!/^[0-9a-fA-F]{4}$/.test(hexDigits)) this.fail("bad \\u escape");
            out += String.fromCharCode(parseInt(hexDigits, 16));
            this.pos += 4;
            break;
          }
          default:
            this.fail(`bad escape \\${e}`);
        }
        continue;
      }
      if (CONTROL_CHARS.test(c)) this.fail("raw control character in string");
      out += c;
      this.pos++;
    }
  }

  // Refuses anything not lexically an integer. Python's json.loads("1e2") is
  // the float 100.0 and json.dumps gives "100.0", so an exponent or fraction
  // may never reach the canonical form. An integer token beyond ±(2^53 − 1)
  // is refused outright rather than silently rounded.
  parseNumber() {
    const rest = this.text.slice(this.pos);
    const match = rest.match(/^-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?/);
    if (!match) this.fail("not a JSON value");
    const [token, , fraction, exponent] = match;
    this.pos += token.length;
    if (fraction !== undefined || exponent !== undefined) {
      throw new CanonicalisationError("no floats in a hashed or signed object");
    }
    const value = Number(token);
    if (!Number.isSafeInteger(value)) {
      throw new CanonicalisationError("integers must stay within ±(2^53 − 1)");
    }
    return value;
  }
}

/**
 * Strict parse of JSON text, then canonicalise. A duplicate key at any depth,
 * a float (including exponent notation such as 1e2) or an integer outside
 * ±(2^53 − 1) is refused — this is the entry point the T01 rejection vectors
 * run through.
 * @param {string} text
 * @returns {Uint8Array}
 */
export function canonicalizeJson(text) {
  if (typeof text !== "string") {
    throw new CanonicalisationError("canonicalizeJson takes JSON text");
  }
  const parser = new JsonParser(text);
  const value = parser.parseValue();
  parser.skipWs();
  if (!parser.atEnd()) parser.fail("trailing characters after JSON value");
  return canonicalize(value);
}

File: app/src/main/java/za/co/vuka/app/api/Canonical.kt
package za.co.vuka.app.api

/**
 * The canonical JSON form of docs/VUKA-2-SPEC.md §5, for the native app: the
 * UTF-8 of Python `json.dumps(obj, sort_keys=True, separators=(",", ":"),
 * ensure_ascii=True)`, byte for byte. A line-for-line port of
 * shared/canonical.js.
 *
 * PROPOSED DEVIATION: §5 rule 5 says "There is no Kotlin canonicaliser". The
 * native app has no JavaScript runtime, so it needs this one to talk to the
 * server. It is held to the same golden and rejection vectors
 * (contracts/vectors/canonical.json) by CanonicalTest; the contract owners
 * must accept it (recorded in the build log).
 *
 * Values: Map<String, Any?> (keys ASCII), List<Any?>, String, Int/Long,
 * Boolean, null. Floats and out-of-range integers are refused.
 */
object Canonical {

    class CanonicalisationError(message: String) : IllegalArgumentException(message)

    private const val MAX_SAFE = 9_007_199_254_740_991L

    fun json(value: Any?): String = StringBuilder().also { encode(value, "$", it) }.toString()

    fun bytes(value: Any?): ByteArray = json(value).toByteArray(Charsets.UTF_8)

    /** Strict parse (duplicate keys, floats, unsafe integers refused), then canonicalise. */
    fun fromJsonText(text: String): String {
        val p = Parser(text)
        val v = p.value()
        p.ws()
        if (!p.atEnd()) throw CanonicalisationError("trailing characters after JSON value")
        return json(v)
    }

    /** Strict parse only: the values this module encodes. */
    fun parse(text: String): Any? {
        val p = Parser(text)
        val v = p.value()
        p.ws()
        if (!p.atEnd()) throw CanonicalisationError("trailing characters after JSON value")
        return v
    }

    private fun encode(value: Any?, path: String, out: StringBuilder) {
        when (value) {
            null -> out.append("null")
            is Boolean -> out.append(if (value) "true" else "false")
            is Int -> out.append(value.toString())
            is Long -> {
                if (value > MAX_SAFE || value < -MAX_SAFE) throw CanonicalisationError("integers must stay within ±(2^53 − 1) (at $path)")
                out.append(value.toString())
            }
            is Float, is Double -> throw CanonicalisationError("no floats in a hashed or signed object (at $path)")
            is String -> escape(value, out)
            is List<*> -> {
                out.append('[')
                value.forEachIndexed { i, item ->
                    if (i > 0) out.append(',')
                    encode(item, "$path[$i]", out)
                }
                out.append(']')
            }
            is Map<*, *> -> {
                val keys = value.keys.map {
                    val k = it as? String ?: throw CanonicalisationError("object keys must be strings (at $path)")
                    if (k.any { c -> c.code > 0x7f }) throw CanonicalisationError("object keys must be ASCII only (at $path)")
                    k
                }.sorted() // ASCII, so code-unit order is code-point order
                out.append('{')
                keys.forEachIndexed { i, k ->
                    if (i > 0) out.append(',')
                    escape(k, out)
                    out.append(':')
                    encode(value[k], "$path.$k", out)
                }
                out.append('}')
            }
            else -> throw CanonicalisationError("type ${value::class.simpleName} is not canonical (at $path)")
        }
    }

    // Python's ESCAPE_ASCII is ([\\"]|[^ -~]): printable ASCII stays literal, the
    // rest escapes. Kotlin strings are UTF-16, so astral characters come out as
    // lowercase surrogate pairs, exactly as Python emits them.
    private fun escape(s: String, out: StringBuilder) {
        out.append('"')
        for (c in s) {
            when {
                c == '"' -> out.append("\\\"")
                c == '\\' -> out.append("\\\\")
                c == '\b' -> out.append("\\b")
                c == '\t' -> out.append("\\t")
                c == '\n' -> out.append("\\n")
                c == '\u000c' -> out.append("\\f")
                c == '\r' -> out.append("\\r")
                c.code in 0x20..0x7e -> out.append(c)
                else -> out.append("\\u").append(c.code.toString(16).padStart(4, '0'))
            }
        }
        out.append('"')
    }

    private class Parser(val text: String) {
        var pos = 0
        fun atEnd() = pos >= text.length
        fun fail(msg: String): Nothing = throw CanonicalisationError("$msg at $pos")
        fun ws() {
            while (!atEnd() && text[pos] in " \t\n\r") pos++
        }

        fun expect(lit: String) {
            if (!text.startsWith(lit, pos)) fail("expected $lit")
            pos += lit.length
        }

        fun value(): Any? {
            ws()
            if (atEnd()) fail("unexpected end of JSON")
            return when (text[pos]) {
                '{' -> obj()
                '[' -> arr()
                '"' -> str()
                't' -> { expect("true"); true }
                'f' -> { expect("false"); false }
                'n' -> { expect("null"); null }
                else -> num()
            }
        }

        fun obj(): Map<String, Any?> {
            expect("{")
            val m = LinkedHashMap<String, Any?>()
            ws()
            if (!atEnd() && text[pos] == '}') { pos++; return m }
            while (true) {
                ws()
                if (atEnd() || text[pos] != '"') fail("expected a string key")
                val k = str()
                if (m.containsKey(k)) throw CanonicalisationError("duplicate key \"$k\" in JSON text")
                ws()
                if (atEnd() || text[pos] != ':') fail("expected ':'")
                pos++
                m[k] = value()
                ws()
                if (atEnd()) fail("expected ',' or '}'")
                when (text[pos]) {
                    ',' -> pos++
                    '}' -> { pos++; return m }
                    else -> fail("expected ',' or '}'")
                }
            }
        }

        fun arr(): List<Any?> {
            expect("[")
            val a = ArrayList<Any?>()
            ws()
            if (!atEnd() && text[pos] == ']') { pos++; return a }
            while (true) {
                a.add(value())
                ws()
                if (atEnd()) fail("expected ',' or ']'")
                when (text[pos]) {
                    ',' -> pos++
                    ']' -> { pos++; return a }
                    else -> fail("expected ',' or ']'")
                }
            }
        }

        fun str(): String {
            pos++ // opening quote
            val sb = StringBuilder()
            while (true) {
                if (atEnd()) fail("unterminated string")
                val c = text[pos]
                when {
                    c == '"' -> { pos++; return sb.toString() }
                    c == '\\' -> {
                        pos++
                        if (atEnd()) fail("unterminated escape")
                        when (val e = text[pos++]) {
                            '"' -> sb.append('"')
                            '\\' -> sb.append('\\')
                            '/' -> sb.append('/')
                            'b' -> sb.append('\b')
                            'f' -> sb.append('\u000c')
                            'n' -> sb.append('\n')
                            'r' -> sb.append('\r')
                            't' -> sb.append('\t')
                            'u' -> {
                                if (pos + 4 > text.length) fail("incomplete \\u escape")
                                val h = text.substring(pos, pos + 4)
                                if (!h.all { it in "0123456789abcdefABCDEF" }) fail("bad \\u escape")
                                sb.append(h.toInt(16).toChar())
                                pos += 4
                            }
                            else -> fail("bad escape \\$e")
                        }
                    }
                    c.code < 0x20 -> fail("raw control character in string")
                    else -> { sb.append(c); pos++ }
                }
            }
        }

        // Anything not lexically an integer is refused: Python reads 1e2 as a float.
        fun num(): Long {
            val m = Regex("^-?(0|[1-9][0-9]*)(\\.[0-9]+)?([eE][+-]?[0-9]+)?").find(text.substring(pos)) ?: fail("not a JSON value")
            pos += m.value.length
            if (m.groupValues[2].isNotEmpty() || m.groupValues[3].isNotEmpty()) {
                throw CanonicalisationError("no floats in a hashed or signed object")
            }
            val v = m.value.toBigInteger()
            if (v > MAX_SAFE.toBigInteger() || v < (-MAX_SAFE).toBigInteger()) {
                throw CanonicalisationError("integers must stay within ±(2^53 − 1)")
            }
            return v.toLong()
        }
    }
}
